data.md

```diff
@@ -955,5 +1007,154 @@ Decoding
+      requires #asInteger( LEAFPATH ) ==Int #asInteger( PATH )
+
+    rule MerkleUpdate ( MerkleLeaf ( LEAFPATH, LEAFVALUE ), PATH, VALUE )
+      => MerkleUpdate ( MerkleUpdate ( .MerkleBranch, LEAFPATH, LEAFVALUE ), PATH, VALUE )
+      requires #sizeByteArray( LEAFPATH ) >Int 0
+       andBool #sizeByteArray( PATH ) >Int 0
+       andBool LEAFPATH[0] =/=Int PATH[0]
+
+    rule MerkleUpdate ( MerkleLeaf ( LEAFPATH, LEAFVALUE ), PATH, VALUE )
+      => #merkleExtensionBuilder( .ByteArray, LEAFPATH, LEAFVALUE, PATH, VALUE ) [owise]
+
+    rule MerkleUpdate ( MerkleExtension ( EXTPATH, EXTTREE ), PATH, VALUE )
+      => MerkleExtension ( EXTPATH, MerkleUpdate ( EXTTREE, .ByteArray, VALUE ) )
+      requires #asInteger( EXTPATH ) ==Int #asInteger( PATH )
+
+    rule MerkleUpdate ( MerkleExtension ( EXTPATH, EXTTREE ), PATH, VALUE )
+      => #merkleExtensionBrancher( MerkleUpdate( .MerkleBranch, PATH, VALUE ), EXTPATH, EXTTREE )
+      requires #sizeByteArray( EXTPATH ) >Int 0
+       andBool #sizeByteArray( PATH ) >Int 0
+       andBool EXTPATH[0] =/=Int PATH[0]
+
+    rule MerkleUpdate ( MerkleExtension ( EXTPATH, EXTTREE ), PATH, VALUE )
+      => #merkleExtensionSplitter( .ByteArray, EXTPATH, EXTTREE, PATH, VALUE ) [owise]
+
+    rule MerkleUpdate ( MerkleBranch( M, _ ), PATH, VALUE )
+      => MerkleBranch( M, VALUE )
+      requires #sizeByteArray( PATH ) ==Int 0
+
+    rule MerkleUpdate ( MerkleBranch( M, BRANCHVALUE ), PATH, VALUE )
+      => #merkleBrancher ( M, BRANCHVALUE, PATH[0], PATH[1 .. #sizeByteArray(PATH) -Int 1], VALUE ) [owise]
+```
+
+Merkle Tree Aux Functions
+-------------------------
+
+```k
+    syntax ByteArray ::= #nibbleize ( ByteArray ) [function]
+                       | #byteify   ( ByteArray ) [function]
+ // --------------------------------------------------------
+    rule #nibbleize ( B ) =>    #asByteStack ( ( B [ 0 ] /Int 16 ) *Int 256 +Int ( B [ 0 ] %Int 16 ) )[0 .. 2]
+                             ++ #nibbleize ( B[1 .. #sizeByteArray(B) -Int 1] )
+      requires #sizeByteArray( B ) >Int 0
+
+    rule #nibbleize ( _ ) => .ByteArray [owise]
+
+    rule #byteify ( B ) =>    #asByteStack ( B[0] *Int 16 +Int B[1] )
+                           ++ #byteify ( B[2 .. #sizeByteArray(B) -Int 2] )
+      requires #sizeByteArray(B) >Int 0
+
+    rule #byteify ( _ ) => .ByteArray [owise]
+
+    syntax ByteArray ::= #HPEncode ( ByteArray, Int ) [function]
+ // ------------------------------------------------------------
+    rule #HPEncode ( X, T ) => #asByteStack ( ( f(T) +Int 1 ) *Int 16 +Int X[0] ) ++ #byteify( X[1 .. #sizeByteArray(X) -Int 1] )
+      requires #sizeByteArray(X) %Int 2 =/=Int 0
+
+    rule #HPEncode ( X, T ) => #asByteStack ( f(T) *Int 16 )[0 .. 1] ++ #byteify( X ) [owise]
+
+    syntax Int ::= f ( Int ) [function,klabel(HPEncodeAux)]
+ // -------------------------------------------------------
+    rule f ( X ) => 0 requires X ==Int 0
+    rule f ( _ ) => 2 [owise]
```

Comment:
Let's avoid calling functions by single letters, why not just `syntax Int ::= HPEncodeAux ( Int ) [function]` directly?